### PR TITLE
Add a join task, passing parent_uid to holder

### PR DESCRIPTION
**Pull request by gdraynz**

### CODE REVIEW

File: tukio/task/join.py [hunk 1]
```diff
@@ -1,47 +1,51 @@
...
-    A join task is a regular task that awaits multiple parents calls using data_received.
-    Task can be overriden for custom behaviours.
-    and decide wether or not unlock the task, wait depending on the calls received
+    A join task is an almost-standard task that allows awaiting for multiple
+    parents and gathering their results using `data_received()`.
+    The `wait_for` config parameter is mandatory and is a list of task IDs.
     """
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
```
Comment (lhavel) [inline tukio/task/join.py]: Bad coding pattern. Never pass arbitrary arguments when you know there's only one possible (kwargs `config`)

File: tukio/task/join.py [hunk 2]
```diff
@@ -1,47 +1,51 @@
...
+        self.data_stash = {}
+        self.wait_for = self.config['wait_for']
+        self.timeout = self.config.get('timeout')
 
     async def execute(self, data):
-        log.debug('join task {} started.'.format(self))
-        self.data_received(data, from_parent=True)
-        await self.unlock
-        log.debug('join task {} done.'.format(self))
+        log.info(
```
Comment (lhavel) [inline tukio/task/join.py]: `DEBUG` instead of `INFO`? We should keep our logs lightweight for production.

File: tukio/task/join.py [hunk 3]
```diff
@@ -1,47 +1,51 @@
...
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.unlock = asyncio.Future()
-        self.data_stash = []
-
-    def ready(self):
-        return len(self.data_stash) >= self.config.get('await_parents', 0)
+        self.unlock = asyncio.Event()
+        self.data_stash = {}
+        self.wait_for = self.config['wait_for']
```
Comment (lhavel) [inline tukio/task/join.py]: can be confusing with the `asyncio.wait_for` function. What's the config dict for that task? Can you give an example?

File: tukio/task/join.py [hunk 4]
```diff
@@ -1,47 +1,51 @@
...
             return
-        log.debug('Received join task data {}'.format(data))
-        self.data_stash.append(data)
-        if self.ready():
-            try:
-                self.unlock.set_result('Done')
-            except InvalidStateError:
-                # Task may have already been unlocked
-                pass
+        log.info("Parent '%s' joined", from_parent)
```
Comment (lhavel) [inline tukio/task/join.py]: Too verbose for production I think

File: tukio/task/join.py [hunk 5]
```diff
@@ -1,47 +1,51 @@
...
+            'Join task waiting for %s (timeout: %s)',
+            self.wait_for,
+            self.timeout
+        )
+        await asyncio.wait_for(self.unlock.wait(), self.timeout)
+        log.info('All parents joined: %s', self.wait_for)
         return self.data_stash
 
-    def data_received(self, data, from_parent=False):
+    def data_received(self, data, from_parent=None):
```
Comment (lhavel) [inline tukio/task/join.py]: As discussed yesterday, this code should run within the context of the initial asyncio Task. But that shall be fixed soon in a separate PR.

File: tukio/task/template.py [hunk 6]
```diff
@@ -27,11 +27,11 @@ def __init__(self, name, uid=None, config=None, topics=None):
     def listen(self):
         return Listen.get(self.topics)
 
-    def new_task(self, *args, loop=None):
+    def new_task(self, *args, loop=None, **kwargs):
```
Comment (lhavel) [inline tukio/task/template.py]: Why didn't you add `from_parent` instead of `**kwargs`. Let's keep the code understandable...

File: tukio/workflow.py [hunk 7]
```diff
@@ -549,12 +549,17 @@ def _run_next_tasks(self, task_tmpl, task):
...
                 if next_task:
-                    joined = self._join_task(next_task, result)
+                    joined = self._join_task(next_task, task_tmpl.uid, result)
                     if not joined:
                         break
                 # Create new task
                 else:
                     next_task = self._new_task(tmpl, result)
+                    # If is a join task, immediately join task's result
+                    if tmpl.name == 'join':
```
Comment (lhavel) [inline tukio/workflow.py]: That way, another developer would have to inherit from ` JoinTask` to code his own or tweak the code of `workflow.py`.
This is not a good idea, we kept it agnostic to avoid such situations. You should find a way to remove those 5 lines in the code.

File: tukio/task/join.py (hunk 3 above)
Comment (gdraynz) [inline tukio/task/join.py]: I find it on the contrary pretty much straightforward as it says exactly what it is doing : wait for these tasks to be done.
Anyway the user that configures the task have no idea what `asyncio.wait_for` is.

File: tukio/task/join.py (hunk 2 above)
Comment (gdraynz) [inline tukio/task/join.py]: I would let this specific log on `INFO` to know that this specific task has started. I'll pass all other logs on `DEBUG`

Comment (lhavel) [inline tukio/task/join.py]: Ok we shall review the logs from workflow executions later to ensure consistency across all tasks

File: tukio/workflow.py [hunk 8]
```diff
@@ -549,12 +549,17 @@ def _run_next_tasks(self, task_tmpl, task):
...
                 # Downstream task already running, join it!
                 if next_task:
-                    joined = self._join_task(next_task, result)
+                    joined = self._join_task(next_task, task_tmpl.uid, result)
                     if not joined:
                         break
                 # Create new task
                 else:
                     next_task = self._new_task(tmpl, result)
+                    # If is a join task, immediately join task's result
```
Comment (lhavel) [inline tukio/workflow.py]: Well, this is actually not the way we should do it.
The task is passed `result` at startup. This is far enough to do the job. Let's keep it simple and don't call `data_received` here. Refactor the `join` task accordingly to stash the 1st result from within `execute`.

File: tukio/task/join.py [hunk 9]
```diff
@@ -17,36 +17,41 @@ class JoinTask(TaskHolder):
...
-    IS_JOIN_TASK = True
-
     def __init__(self, config):
         super().__init__(config)
-        self.unlock = asyncio.Event()
-        self.data_stash = {}
-        self.wait_for = self.config['wait_for']
-        self.timeout = self.config.get('timeout')
+        self._unlock = asyncio.Event()
+        self._data_stash = []
```
Comment (gdraynz) [inline tukio/task/join.py]: Shouldn't we use a dict to keep the task_id/task_result pairs ? to be able to use these results more precisely within next tasks
